### DISCUSSION TITLE
api.assumption is not a function

### DISCUSSION THREAD

**Discussion by derroobin** [babel/babel#12949] [General]
Hi,
i get an error, while building my package: `(babel plugin) TypeError: [BABEL] src/index.tsx: api.assumption is not a function (While processing: "node_modules/microbundle-crl/node_modules/@babel/preset-env/lib/index.js$3")`

the error seems to be in: node_modules/@babel/plugin-proposal-optional-chaining/lib/index.js:71

```javascript
const noDocumentAll = (_api$assumption = api.assumption("noDocumentAll")) != null ? _api$assumption : loose;
const pureGetters = (_api$assumption2 = api.assumption("pureGetters")) != null ? _api$assumption2 : loose;
```

is api.assumption supposed to be defined or is this an error in plugin-proposal-optional-chaining?

**Comment by nicolo-ribaudo** [comment]
Can you share the full stack trace, and the versions of `@babel/preset-env`, `@babel/plugin-proposal-optional-chaining`, `@babel/helper-plugin-utils` and `@babel/core`?

**Reply by rdsedmundo** [reply]
I'm also facing this on the latest version of Babel while using create-react-app, after I upgraded Babel dependencies.

```
    "@babel/cli": "^7.13.0",
    "@babel/core": "^7.13.8",
    "@babel/plugin-proposal-class-properties": "^7.13.0",
    "@babel/plugin-transform-typescript": "^7.13.0",
    "@babel/preset-env": "^7.13.9",
    "babel-loader": "^8.2.2",
    "react-scripts": "^4.0.3"
```


```
TypeError: [BABEL] /home/runner/work/redacted/redacted/packages/web-app/src/index.js: api.assumption is not a function (While processing: "/home/runner/work/redacted/redacted/node_modules/babel-preset-react-app/index.js$0$2")
    at Generator.next (<anonymous>)
    at Generator.next (<anonymous>)
    at Generator.next (<anonymous>)
    at cachedFunction.next (<anonymous>)
```

**Reply by hnhegde** [reply]
I encountered this error on AWS Amplify. 
```
2021-03-16T16:36:10.380Z [INFO]: Creating an optimized production build...
2021-03-16T16:36:11.999Z [INFO]: Failed to compile.
2021-03-16T16:36:12.000Z [INFO]: ./src/index.tsx
                                 TypeError: [BABEL] /codebuild/output/src772776919/src/square-deal-ui/src/index.tsx: api.assumption is not a function (While processing: "/codebuild/output/src772776919/src/square-deal-ui/node_modules/babel-preset-react-app/index.js$0$2")
                                 at Generator.next (<anonymous>)
                                 at Generator.next (<anonymous>)
                                 at Generator.next (<anonymous>)
                                 at cachedFunction.next (<anonymous>)
2021-03-16T16:36:12.019Z [WARNING]: error Command failed with exit code 1.
2021-03-16T16:36:12.020Z [INFO]: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
2021-03-16T16:36:12.026Z [ERROR]: !!! Build failed
2021-03-16T16:36:12.026Z [ERROR]: !!! Non-Zero Exit Code detected
```
Attached herewith is the complete log. 
[log (1).txt](https://github.com/babel/babel/files/6150775/log.1.txt)

I tried adding `yarn cache clean --all ` to the build yaml file. But didn't help.

**Reply by hnhegde** [reply]
Root cause - not identified. Solution - brute force. 
I destroyed the Amplify app and recreated it. Compilation succeeded and the build also succeeded.

**Reply by emutanen** [reply]
I got the same error when upgrading `react-scripts` to 4.0.3. 
I am using CRA with typescript 4.3.4 and Semantic UI.

**Comment by rdsedmundo** [comment]
I cleared my Babel and Yarn cache, did a reinstall, and it started to work again. This is not the first time that this happens between Babel upgrades, although it was a different error at the time. It's weird that minor/patch upgrades keep introducing those inconsistencies in the cache, maybe it should be invalidated once a new version is installed.

**Reply by nicolo-ribaudo** [reply]
It's probably caused by an incompatibility between two different versions of two different packages.

This kind of bugs is quite hard to predict, it's almost impossible to reproduce without a repository that shows the bug, and it's quite easy to fix on the user side (as you did, by re-installing everything).

It's a bit unfortunate, but usually we try to fix them anyway to ensure maximum compatibility.

**Comment by areyes-ut** [comment]
I managed to get past `api.assumption is not a function` by `rm -rf node_modules` and re-installing packages (whether with `yarn` or `npm`).

**Reply by nicolo-ribaudo** [reply]
Do you have the old lockfile by chance? It might help understand why this error is happening.

**Reply by areyes-ut** [reply]
I do, but the codebase is proprietary and closed source.  Sorry.  :\

**Comment by nad182** [comment]
Had a similar issue with CircleCI and Netlify using cached files during builds after upgrading `@babel/preset-env` from `7.9.5` to `7.13.9` (dependabot). Manually changing `key` in `restore_cache`  instruction in CircleCI `config.yml` and restarting Netlify with cleared cached worked like a charm. Happy, I found this discussion!

**Comment by sturmenta** [comment]
using babel in `react-native` my solution to `api.assumption is not a function` was upgrading everything related to babel

<img width="1137" alt="Screen Shot 2021-05-20 at 13 13 18" src="https://user-images.githubusercontent.com/30802967/119014102-05378d80-b96e-11eb-881e-833658812658.png">

**Comment by pavlos256** [comment]
I got this too. Very strange. I did `rm -rf node_modules/`, then `yarn` and now it works. I have a lockfile!

**Reply by MiriamOfir** [reply]
thanks, that's helped me

**Comment by emutanen** [comment]
I have two lockfiles in the src folder, ```yarn.lock``` and ```package-lock.json```. Removing the package-lock file, cleaning cache with yarn and reinstalling node_modules with yarn helped to remove this error. If anyone knows what two different lockfiles can cause then it would be interesting to understand how is this problem surfacing in such a setup where devs are moving from npm to yarn.

**Reply by nicolo-ribaudo** [reply]
One of those lockfile is ignored, depending on which package manager you use. If you move from npm to yarn, you should delete `package-lock.json`.

**Comment by yousufdarr** [comment]
For me reinstalling the plugin that had the issue i.e. `@babel/plugin-proposal-private-methods` solved the problem.

**Comment by pierregambaud** [comment]
Adding the following to my `package.json` solved my "api.addExternalDependency is not a function" error:
```
"resolutions": {
  "@babel/core": "^7.20.2",
  "babel-loader": "8.3.0"
}
```

_Source: https://dev.to/goatandsheep/how-to-use-babels-new-addexternaldependency-api-338h_
